### PR TITLE
[DEV-3055] Route table

### PR DESCRIPTION
**Pull request by Yannick1712**
#### Pre-Release

- [ ] Check migrations
  - No database related infos (sqldb-xxx)
  - Impact on GS (new/removed columns)
- [ ] Check for linter errors (in PR)
- [ ] Test basic user operations (on [DFX services](https://dev.services.dfx.swiss))
  - Login/logout
  - Buy/sell payment request
  - KYC page

#### Post-Release

- Test basic user operations
- Monitor application insights log

### CODE REVIEW

**Comment by Yannick1712** [comment]
- [ ] add migration

File: src/subdomains/core/route/route.controller.ts
Comment (davidleomay) [inline src/subdomains/core/route/route.controller.ts:1]: There is already a `RouteController`, but it's a bit misplaced 🤔

File: src/subdomains/core/route/route.controller.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+export class RouteController {
+  constructor(private readonly routeService: RouteService) {}
+
+  @Post()
+  @ApiBearerAuth()
+  @UseGuards(AuthGuard(), new RoleGuard(UserRole.ADMIN))
+  @ApiExcludeEndpoint()
+  async createRoute(@Body() dto: CreateRouteDto): Promise<Route> {
+    return this.routeService.createRoute(dto);
+  }
```
[the remark this answers is not in the document]
Comment (Yannick1712) [reply src/subdomains/core/route/route.controller.ts:23]: Yeah if label is nullable true thats the best way 👍

File: src/subdomains/core/route/route.controller.ts
Comment (Yannick1712) [reply src/subdomains/core/route/route.controller.ts:1]: Hm yeah 🤔 what should we do about it? Who is using the old endpoint? If e.g. only services uses it, we could rename it to depositRoute controller 👍

Comment (davidleomay) [reply src/subdomains/core/route/route.controller.ts:1]: Can't we move the existing code into the new controller?

Comment (Yannick1712) [reply src/subdomains/core/route/route.controller.ts:1]: Haha ah yeah true 👍

File: src/subdomains/core/route/route.entity.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import { User } from 'src/subdomains/generic/user/models/user/user.entity';
+import { Column, Entity, OneToOne } from 'typeorm';
+import { Buy } from '../buy-crypto/routes/buy/buy.entity';
+import { Swap } from '../buy-crypto/routes/swap/swap.entity';
+import { Sell } from '../sell-crypto/route/sell.entity';
+
+@Entity()
+export class Route extends IEntity {
+  @Column({ length: 256, unique: true, nullable: true })
+  label: string;
```
Comment (davidleomay) [inline src/subdomains/core/route/route.entity.ts:12]: I think this unique index is not helpful/required?

File: src/subdomains/core/route/route.service.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    const entity = await this.routeRepo.findOne({
+      where: { id },
+      relations: { buy: true, sell: true, swap: true },
+    });
+    if (!entity) throw new NotFoundException('Route not found');
+
+    const update = this.routeRepo.create(dto);
+
+    if (update.label && (await this.routeRepo.existsBy({ label: update.label })))
+      throw new BadRequestException('Label already in use');
```
Comment (davidleomay) [inline src/subdomains/core/route/route.service.ts:83]: Same here

File: src/subdomains/core/route/route.service.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+    if (dto.sell) {
+      entity.sell = await this.sellService.get(undefined, dto.sell.id);
+      if (!entity.sell) throw new NotFoundException('Sell not found');
+    }
+
+    if (dto.swap) {
+      entity.swap = await this.swapService.get(undefined, dto.swap.id);
+      if (!entity.swap) throw new NotFoundException('Swap not found');
+    }
```
Comment (davidleomay) [inline src/subdomains/core/route/route.service.ts:68]: I think these checks can be removed. The method is only used internally with an existing route.

File: src/subdomains/core/route/route.service.ts
Comment (davidleomay) [inline src/subdomains/core/route/route.service.ts:1]: I think we should add some code create a route directly when creating a buy/sell/swap. Or do you want to add this later?

File: src/subdomains/core/route/route.entity.ts (hunk 2 above)
Comment (Yannick1712) [reply src/subdomains/core/route/route.entity.ts:12]: That's what the ticket says or? 🤔 
Robin said in the ticket, the label col has to be unique on the complete table

File: src/subdomains/core/route/route.service.ts
Comment (Yannick1712) [reply src/subdomains/core/route/route.service.ts:1]: Yes right, thats a second step. First sync all existing routes

File: src/subdomains/core/route/route.entity.ts (hunk 2 above)
Comment (davidleomay) [reply src/subdomains/core/route/route.entity.ts:12]: Ah, true, you're right

File: src/subdomains/core/route/route.service.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+      }
+    }
+  }
+
+  async createRoute(dto: CreateRouteDto): Promise<Route> {
+    const entity = this.routeRepo.create(dto);
+
+    if (dto.buy) entity.buy = await this.buyService.get(undefined, dto.buy.id);
+    if (dto.sell) entity.sell = await this.sellService.get(undefined, dto.sell.id);
+    if (dto.swap) entity.swap = await this.swapService.get(undefined, dto.swap.id);
```
Comment (davidleomay) [inline src/subdomains/core/route/route.service.ts:57]: Remove

**Review by davidleomay — APPROVED** [review]

**Pull request merged**